examples/recurrent_attention_machine_translation.py

```diff
@@ -0,0 +1,962 @@
+                       attended,
+                       attention_states,
+                       attended_mask,
+                       training=None):
+        # there must be two attended sequences (verified in build)
+        [attended, u] = attended
+        attended_mask = attended_mask[0]
+        h_cell_tm1 = cell_states[0]
+
+        # compute attention weights
+        w = K.repeat(K.dot(h_cell_tm1, self.W_a) + self.b_UW, K.shape(attended)[1])
+        e = K.exp(K.dot(K.tanh(w + u), self.v_a) + self.b_v)
+
+        if attended_mask is not None:
+            e = e * K.cast(K.expand_dims(attended_mask, -1), K.dtype(e))
+
+        # weighted average of attended
+        a = e / K.sum(e, axis=1, keepdims=True)
+        c = K.sum(a * attended, axis=1, keepdims=False)
+
+        return c, [c]
+
+    def attention_build(self, input_shape, cell_state_size, attended_shape):
+        if not len(attended_shape) == 2:
+            raise ValueError('There must be two attended tensors')
+        for a in attended_shape:
+            if not len(a) == 3:
+                raise ValueError('only support attending tensors with dim=3')
+        [attended_shape, u_shape] = attended_shape
+
+        # NOTE _attention_size must always be set in `attention_build`
+        self._attention_size = attended_shape[-1]
+        units = u_shape[-1]
+
+        kernel_kwargs = dict(initializer=self.kernel_initializer,
+                             regularizer=self.kernel_regularizer,
+                             constraint=self.kernel_constraint)
+        self.W_a = self.add_weight(shape=(cell_state_size[0], units),
+                                   name='W_a', **kernel_kwargs)
+        self.v_a = self.add_weight(shape=(units, 1),
+                                   name='v_a', **kernel_kwargs)
+
+        bias_kwargs = dict(initializer=self.bias_initializer,
+                           regularizer=self.bias_regularizer,
+                           constraint=self.bias_constraint)
+        self.b_UW = self.add_weight(shape=(units,),
+                                    name="b_UW", **bias_kwargs)
+        self.b_v = self.add_weight(shape=(1,),
+                                   name="b_v", **bias_kwargs)
+
+    def get_config(self):
+        config = {
+            'kernel_initializer': initializers.serialize(self.kernel_initializer),
+            'bias_initializer': initializers.serialize(self.bias_initializer),
+            'kernel_regularizer': regularizers.serialize(self.kernel_regularizer),
+            'bias_regularizer': regularizers.serialize(self.bias_regularizer),
+            'kernel_constraint': constraints.serialize(self.kernel_constraint),
+            'bias_constraint': constraints.serialize(self.bias_constraint)
+        }
+        base_config = super(DenseAnnotationAttention, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+if __name__ == '__main__':
+    DATA_DIR = 'data/wmt16_mmt'
```

Comment:
Can I suggest the following:
```python
    from keras.utils.data_utils import get_file
    base_name = 'wmt16_mmt_'
    origin = 'http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/'
    get_file(base_name + 'train', origin=origin + 'training.tar.gz', untar=True)
    get_file(base_name + 'val', origin=origin + 'validation.tar.gz', untar=True)
    tar_file = get_file(base_name + 'test',
                        origin=origin + 'mmt16_task1_test.tar.gz',
                        untar=True)

    DATA_DIR = os.path.dirname(tar_file)
```

Taking "standalone" a step further.

Reply:
Sure :D

Reply:
Since the extracted files won't have `base_name` I create a new `cache_subdir` instead - otherwise files with a very generic name (`train.en`) ends up in the `datasets` dir.

Reply:
Good idea, I didn't find a way to make a subdir (I didn't look much into it).